packages/open-truss/src/components/OTUqiDataProvider.tsx

```diff
@@ -0,0 +1,134 @@
+import React from 'react'
+import {
+  withChildren,
+  BaseOpenTrussComponentV1PropsShape,
+  type BaseOpenTrussComponentV1,
+} from '../configuration/engine-v1'
+import { type UqiColumn } from '../uqi/uqi'
+import { type Signal, StringSignal } from '../signals'
+import { z } from 'zod'
+
+interface SynchronousQueryResult {
+  rows: SynchronousQueryRow[]
+  metadata: SynchronousQueryMetadata
+}
+
+interface SynchronousQueryRow {
+  values: SynchronousQueryValue[]
+}
+
+interface SynchronousQueryValue {
+  key: string
+  type: string
+  value: string
+}
+
+interface SynchronousQueryMetadata {
+  columns: SynchronousQueryColumn[]
+}
+
+interface SynchronousQueryColumn {
+  name: string
+  type: string
+}
+
+export const Props = BaseOpenTrussComponentV1PropsShape.extend({
+  ...withChildren,
+  source: StringSignal,
+  query: StringSignal,
+  output: z.record(
+    z.union([z.record(z.string()), z.array(z.record(z.string()))]),
+  ),
+})
+
+const UqiDataProvider: BaseOpenTrussComponentV1<z.infer<typeof Props>> = (
+  props,
+) => {
+  const { query, children, output, source } = props
+  const [queryResults, setQueryResults] =
+    React.useState<SynchronousQueryResult>()
+
+  React.useEffect(() => {
+    const fetchData = async function (): Promise<undefined> {
+      const result = await fetch('/api/synchronous-uqi-query', {
+        method: 'POST',
+        body: JSON.stringify({ query, source }),
+      })
+      const deserialized = await result.json()
+      setQueryResults(deserialized)
+    }
+
+    fetchData().catch(console.error)
+  }, [])
+
+  if (queryResults === undefined) return <></>
+
+  Object.entries(output).forEach(([k, v]) => {
+    // TODO make sure signals code is putting in signals even if compoennt doesn't declare that as a property
+    const signal = (props as Record<string, any>)[k]
+    if (signal === undefined) {
+      // TODO figure out how we want to handle this
+      console.log('missing signal in data component')
+      return
+    }
+
+    signal as Signal
+
+    const shape = Array.isArray(v) ? v[0] : v
+
+    if (!isShapeValid(shape, queryResults.metadata.columns)) {
+      // TODO handle error better
+      console.log('shape of data from server does not match output of query')
+    }
+
+    const result = Array.isArray(v) ? queryResults.rows : queryResults.rows[0]
+    signal.value = result
+  })
+
+  return children
+}
+
+// validates if the data conforms to the shape provided.
+// shape - key is the name of the property and the value is the javascript type
+// data - json object with followig shape
+//  [
+//    { key: 'greeting',  type: 'String', value: 'hello' },
+//    { key: 'goodbye',  type: 'String', value: 'byebye' }
+//  ]
+//
+function isShapeValid(
```

Comment:
Ooh I wonder if with #129 knowing about each signal's `valueShape`, we could just do a simple zod parse for this! Like `signal.valueShape.parse(queryResult.rows)` (or `.rows[0]`). This would require to have the top level `results: Record` definition be more specific, which is maybe a good thing? I wonder if we could have a way to declare signals completely from the config. Now that I'm getting this far, I realize this might be out of scope of this PR 😸